backend/fms_core/services/experiment_run_info.py

```diff
@@ -0,0 +1,347 @@
+from typing import Dict, Iterable, List, TextIO, Union
+from dataclasses import asdict, dataclass
+import os
+import json
+from django.conf import settings
+from fms_core.utils import make_timestamped_filename
+from fms_core.coordinates import convert_alpha_digit_coord_to_ordinal
+from fms_core.containers import CONTAINER_KIND_SPECS
+from fms_core.models import (
+    Biosample,
+    Container,
+    DerivedSample, 
+    DerivedBySample,
+    ExperimentRun, 
+    Index,
+    Individual,
+    Instrument, 
+    Library,
+    Process,
+    ProcessMeasurement,
+    Project,
+    PropertyValue,
+    Protocol,
+    Sample, 
+    SampleLineage,
+)
+
+
+Obj_Id = Union[int, None]
+
+@dataclass
+class RunInfoSample:
+    # Biosample name, actually
+    pool_name: Union[str, None]
+    sample_name: Union[str, None]
+
+    derived_sample_obj_id: Obj_Id = None
+    biosample_obj_id: Obj_Id = None
+
+    # Flowcell lane containing the sample
+    container_coordinates: Union[str, None] = None
+    lane: Union[int, None] = None
+
+    project_obj_id: Obj_Id = None
+    project_name: Union[str, None] = None
+    hercules_project_id: Union[str, None] = None
+    hercules_project_name: Union[str, None] = None
+
+    pool_volume_ratio: Union[float, None] = None
+
+    # individual fields
+    expected_sex: Union[str, None] = None
+    ncbi_taxon_id: Union[int, None] = None
+    taxon_name: Union[str, None] = None
+    
+    # library fields
+    library_type: Union[str, None] = None
+    library_size: Union[float, None] = None
+    index_set_obj_id: Union[str, None] = None
+    index_set_name: Union[str, None] = None
+    index_obj_id: Obj_Id = None
+    index_name: Union[str, None] = None
+    index_sequence_3_prime: Union[List[str], None] = None
+    index_sequence_5_prime: Union[List[str], None] = None
+    library_kit: Union[str, None] = None
+
+    # capture fields
+    capture_kit: Union[str, None] = None
+    capture_baits: Union[str, None] = None
+
+    #ChIP-Seq
+    chip_seq_mark: Union[str, None] = None
+    
+
+@dataclass
+class RunInfo:
+    # Manifest version (1.0.0 to start) 
+    version: str
+
+    # Experiment run name and id
+    run_name: str
+    run_obj_id: int
+    run_start_date: Union[str, None]
+
+    # Flowcell / container for experiment
+    container_obj_id: int
+    container_barcode: str
+
+    platform: str
+    instrument_serial_number: str
+    instrument_type: str
+
+    samples: List[RunInfoSample]
+
+
+RUN_INFO_FILE_VERSION = "1.0.0"
+
+def generate_run_info_file(experiment_run: ExperimentRun):
+    '''
+    Generate the run info for an experiment, to be submitted to tech dev for run processing.
+
+    This function writes the data to the file it is given. The file must be text-based, open,
+    and writable.
+
+    The output is a RunInfo object in the JSON format.
```

Comment:
There is no return value for this function. If i want the file i must got fetch the file at a pre-established place ... this is likely more risk and more work. There is also no error return for a function that has IO which is not very safe. I would encourage to rework the signature. Maybe take the file location as additional input and return error messages on failure. Another way could be to return the file object to be saved to disk or sent as a response (this could be done in addition to saving it to disk too) failure would return None. Either way it would be more safe and clear.

Reply:
Yeah, I wasn't too sure about this function either. I trashed it and moved all the code into the service, where I think it makes more sense.